### PR TITLE
PLAT-76082: Add SharedStateDecorator

### PR DESCRIPTION
**Pull request by ryanjduffy**
### Checklist

* [ ] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [ ] Documentation was added or is not needed

* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
We have requirements to "remember" the scroll position when navigating forward (positive `index` change) in a Panels stack so it can be restored when navigating back (negative `index` change) to it.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
* Add a new internal HOC, `SharedStateDecorator`, which maintains a private key/value data store and makes it available via a `SharedState` context object
* Add the `SharedStateDecorator` to `Panels` and `Panel` thereby establishing localized data stores for their respective subtree
* Add a handler to `Scroller` to notify its upstream `SharedState` of scroll position changes

The process is as follows:
1. `Panels` creates a new data store (via `SharedStateDecorator`) to track the state of each of its panel instances
2. `Panel` creates its own data store keyed by the `data-index` value passed by `Panels` (via `Viewport`)
3. `Scroller` uses the data store created and passed by `Panel` to store the current scroll position in an `onScroll` handler
4. When navigating forward, new keys are appended to the `Panels` data store for the new Panels. When navigating backward, all keys greater than the current index are deleted thereby erasing their shared state.
5. When returning to a prior `Panel`, `SharedStateDecorator` retrieves the shared state from its parent (`Viewport`). If that `Panel` contains a `Scroller`, the `Scroller` retrieves the scroll position in `componentDidMount` and imperatively scrolls to the cached position.

### CODE REVIEW

**Comment by seunghoh** [comment]
@ryanjduffy Restore scroll logic is in the `scroller` This means VirutalList in the Panel won't support. Is this intentional?

**Comment by ryanjduffy** [comment]
I started with that since VirtualList has its placeholder to help restore scroll position. If we move ahead with this, we'll need to decide how to reconcile those two approaches.

**Comment by seunghoh** [comment]
Overall, I like your approach 😀
One minor concern is `sharedStateId = 'scroller'`. Channel Edit has 3 List components, but we can guide app developer to add prop explicitly. 
It looks like logic in `scroller` can move to `scrollerable` + scroll to index and remove placeholder ?

**Comment by ryanjduffy** [comment]
@germboy, @viodragon2, @Djspaceg, @webOS101 - I've pushed a couple more commits to this and think it's getting close. I've moved the position caching logic into Scrollable, converted to use `id` instead of a new prop, and added a `noSharedState` prop to `SharedStateDecorator` to allow some control.

Outstanding items:
* Determine if we should document `noSharedState` as a public prop on Panels/Panel
* Document `id` prop on Scroller/Panels/Panel
* Document `SharedStateDecorator`
* Test with Virtual(Grid)List(Native) incl async data loading scenarios
* Create a sample for testing
* Unit tests

File: packages/moonstone/Panels/Panels.js [hunk 1]
```diff
@@ -184,22 +208,24 @@ const PanelsBase = kind({
...
 
 		return (
-			<div {...rest}>
+			<div {...rest} id={id}>
 				{applicationCloseButton}
 				<Viewport
 					arranger={arranger}
 					childProps={childProps}
 					generateId={generateId}
+					id={`viewport-${id}`}
```
Comment (ryanjduffy) [inline packages/moonstone/Panels/Panels.js]: Fix this

File: packages/moonstone/Scrollable/Scrollable.js [hunk 2]
```diff
@@ -73,6 +74,8 @@ const configureSpotlightContainer = ({'data-spotlight-id': spotlightId, focusabl
 	});
 };
 
+const isValidId = (id) => id !== '';
```
Comment (ryanjduffy) [inline packages/moonstone/Scrollable/Scrollable.js]: Probably don't need this

File: packages/ui/VirtualList/VirtualListBase.js [hunk 3]
```diff
@@ -276,10 +287,14 @@ const VirtualListBaseFactory = (type) => {
 			this.setContainerSize();
 		}
 
-		componentDidUpdate (prevProps) {
+		componentDidUpdate (prevProps, prevState) {
 			// TODO: remove `this.hasDataSizeChanged` and fix ui/Scrollable*
 			this.hasDataSizeChanged = (prevProps.dataSize !== this.props.dataSize);
 
+			if (prevState.firstIndex !== this.state.firstIndex && this.props.onUpdateItems) {
```
Comment (ryanjduffy) [inline packages/ui/VirtualList/VirtualListBase.js]: remove onUpdateItems condition

File: packages/moonstone/internal/SharedStateDecorator/tests/SharedStateDecorator-specs.js [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+			</Component>
+		);
+
+		const expected = {
+			// "data" is grouped by id so the top level key is the current id
+			outer: {
+				// this is set by value.set('key', 'value');
+				key: 'value',
+				// when a descendant is added, it pushes its data upstream stored by its id
+				inner: {
```
Comment (ryanjduffy) [inline packages/moonstone/internal/SharedStateDecorator/tests/SharedStateDecorator-specs.js]: Might be a problem if `inner` were renamed to an existing `key` value. Need to consider more.

File: packages/moonstone/CHANGELOG.md [hunk 5]
```diff
@@ -16,6 +17,7 @@ The following is a curated list of changes in the Enact moonstone module, newest
 
 - Fonts to correctly use the new font files and updated the international font name from "Moonstone LG Display" to "Moonstone Global"
 - `moonstone/Dropdown` `children` propType so it supports the same format as `ui/Group` (an array of strings or an array of objects with props)
+- `moonstone/FormCheckbox`, `moonstone/Input`, `moonstone/ProgressBar`, `moonstone/RadioItem`, `moonstone/SwitchItem`, and `moonstone/Tooltip` light skin colors.
```
Comment (webOS101) [inline packages/moonstone/CHANGELOG.md]: This looks like a stray

Comment (ryanjduffy) [reply packages/moonstone/CHANGELOG.md]: it was in the wrong spot and i moved it when resolving the merge conflict

**Pull request merged**